internal/cover/cover.go

```diff
@@ -0,0 +1,218 @@
+	return &coverageHelper{
+		covermode,
+		coverappend,
+		fileReader,
+		0,
+		map[int]boundary{},
+		map[int]int{},
+	}
+}
+
+// Annotate annotates program AST with pieces of coverage tracking code
+func (cov *coverageHelper) Annotate(prog *ast.Program) {
+	prog.Begin = cov.annotateStmtsList(prog.Begin)
+	prog.Actions = cov.annotateActions(prog.Actions)
+	prog.End = cov.annotateStmtsList(prog.End)
+	prog.Functions = cov.annotateFunctions(prog.Functions)
+}
+
+// StoreCoverData writes result coverage report data to coverprofile file
+func (cov *coverageHelper) StoreCoverData(coverprofile string, coverData map[string]string) error {
+	// 1a. If file doesn't exist - create and write covermode line
+	// 1b. If file exists and coverappend=true  - open it for writing in append mode
+	// 1c. If file exists and coverappend=false - truncate it and follow 1a.
+	// 2.  Write all coverData lines
+
+	coverDataInts := prepareCoverData(coverData)
+	isNewFile := true
+
+	var f *os.File
+	if _, err := os.Stat(coverprofile); os.IsNotExist(err) {
+		f, err = os.OpenFile(coverprofile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
+		if err != nil {
+			return err
+		}
+	} else if err == nil {
+		// file exists
+		fileOpt := os.O_TRUNC
+		if cov.coverappend {
+			isNewFile = false
+			fileOpt = os.O_APPEND
+		}
+		f, err = os.OpenFile(coverprofile, os.O_WRONLY|fileOpt, 0644)
+		if err != nil {
+			return err
+		}
+	} else {
+		panic(err)
+	}
+
+	if isNewFile {
+		_, err := f.WriteString("mode: " + cov.covermode + "\n")
+		if err != nil {
+			return err
+		}
+	}
+
+	for i := 1; i <= cov.annotationIdx; i++ {
+		_, err := f.WriteString(renderCoverDataLine(cov.boundaries[i], cov.stmtsCnt[i], coverDataInts[i]))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func prepareCoverData(coverData map[string]string) map[int]int {
+	res := map[int]int{}
+	for k, v := range coverData {
+		ki, err := strconv.Atoi(k)
+		if err != nil {
+			panic("non-int index in coverData: " + k)
+		}
+		vi, err := strconv.Atoi(v)
+		if err != nil {
+			panic("non-int value in coverData: " + v)
+		}
+		res[ki] = vi
+	}
+	return res
+}
+
+func (cov *coverageHelper) annotateActions(actions []*ast.Action) (res []*ast.Action) {
+	for _, action := range actions {
+		action.Stmts = cov.annotateStmts(action.Stmts)
+		res = append(res, action)
+	}
+	return
+}
+
+func (cov *coverageHelper) annotateFunctions(functions []*ast.Function) (res []*ast.Function) {
+	for _, function := range functions {
+		function.Body = cov.annotateStmts(function.Body)
+		res = append(res, function)
+	}
+	return
+}
+
+func (cov *coverageHelper) annotateStmtsList(stmtsList []ast.Stmts) (res []ast.Stmts) {
+	for _, stmts := range stmtsList {
+		res = append(res, cov.annotateStmts(stmts))
+	}
+	return
+}
+
+// annotateStmts takes a list of statements and adds counters to the beginning of
+// each basic block at the top level of that list. For instance, given
+//
+//	S1
+//	if cond {
+//		S2
+//	}
+//	S3
```

Comment:
I noticed in the `go tool cover` output from running coverage on `gron.awk`, it doesn't show `if` statement lines as green, bug grey. Shouldn't they be green, because it runs the `if` statement? This seems like a bug. Compared to Go, for example, Go will highlight the `if` line green, regardless of whether the `if` block is executed. That seems right to me, because the condition is always tested.

Similarly, `for` and `while` lines aren't green, but grey, even in a function that's executed.

Reply:
Good catch! Fixed.